File: src/optical_element/MATLAB/w_oe.m
classdef w_oe < handle
    %光学元器件父类 ，所有光学元器件的共同特征
    properties
        beam_in;  % 入射光束对象
        chi;      % 光学元器件法线方向
        ds;       % 物平面 到光学元器件的距离（物距）单位 [m]
        di;       % 光学元器件到 像平面的距离 （像）
        theta;    % 光束掠入射角度 [mrad]
        beam_out; % 出射光束线
        surface;  % 光学元器件面型对象
        g_mirror; % 追迹象光学几何镜子
        gbeam_in;
        grating;
    end    
    
    methods %初始化
        function obj = w_oe( beam_in,ds,di,chi,theta,surface,grating)         
          obj.chi =chi;
          obj.ds= ds;
          obj.di =di;
          obj.theta =theta;
          obj.beam_in =beam_in;  
          obj.surface=surface;  
          obj.grating =grating;
          %*********************↑初始化
        end         
    end   
    
    methods
        function reflect(obj)
              obj.create_g_beam();%产生几何光束              
              obj.tracing();%几何追迹
              g = obj.g_mirror;             
              delta_h = obj.surface.value(g.Z2,g.X2);                 % 计算光学元器件不同位置面型对应的高度 ；
              phase_s = -2*2*pi/obj.beam_in.wavelength.*delta_h.*g.cos_Alpha; % sin(\theta)=-1a.M1;通过面型高度计算产生的相位差 ；
              %-------------------------↑ ↑ ↑ ------------------计算面型误差
              obj.beam_out=obj.creat_w_beam(phase_s);             
              % 将面型误差和光学元器件造成的相位带入到原光束中，形成新的光束；
                  %从波动光束创建几何光束
            % obj.transform_coefficient(g.Cff);
             
        end
    end 
    
    methods % 追迹的光学元器件
        function tracing(obj)
           
        end
    end
    
    methods  %面型畸变
        function  transform_coefficient(obj,Cff)
            if obj.chi ==0 || obj.chi == 180
            obj.beam_out.X = obj.beam_out.X;
            obj.beam_out.Y = obj.beam_out.Y ;
            else
                obj.beam_out.X = obj.beam_out.X;
                obj.beam_out.Y = obj.beam_out.Y ;
            end
        end
    end
    
     methods %从波动光束产生几何光束      
        function create_g_beam(obj)  
         beam= obj.beam_in;
         Phase_field = Furion_phase_unwrap2(beam.field);          % 相位解包裹        
         optical_path= Phase_field/(2*pi)*beam.wavelength;        % 相位转换为光程 单位[m]
         dx = mean(diff(beam.X(1,:)));                            % X 方向的间距  单位[m]
         dy = mean(diff(beam.Y(:,1)));                            % Y方向的间距 单位[m]
         [Fx,Fy] = gradient(optical_path,dx,dy);                 % 求X方向和Y方向的梯度
         theta_x = -atan(Fx);                                    % 由梯度计算角度
         theta_y = -atan(Fy);                                    % 由梯度计算角度                                
         obj.beam_in.phase_field =Phase_field;                           %相位场分布
         
         phi = theta_x(:)';                       % x'    x 方向角度                                       
         psi = theta_y(:)';                       % y'    y 方向角度
         XX = beam.X(:)';                          %网格X坐标转换成散点坐标
         YY =beam. Y(:)';                          %网格Y坐标转换成散点坐标 
         lambda = beam.wavelength;                 % 波动波长和几何波长相同
         obj.gbeam_in = g_beam(XX,YY,phi,psi,lambda);    %产生几何光束
         obj.beam_in.XX =XX;
         obj.beam_in.YY =YY;
        end         
     end
    
       methods   % 构造波动光束 不考虑畸变，将增加的相位直接加在原来的相位对应的位置之上；
        function  wbeam_out=creat_w_beam(obj,s_phase)     % 经过光学元器件后增加的相位（不考虑矩阵畸变）；
       
          
            wbeam=obj.beam_in;
            Phase = obj.g_mirror.Phase + s_phase ;
            phase1 = reshape(Phase,[wbeam.N, wbeam.N]);                             % 将增加的散点相位转化为二维网格矩阵相位      ；
            Field_new = abs(wbeam.field).*exp(1i*(wbeam.phase_field+phase1));       % 通过增加的相位构建新的光的标量复数场分布    ；
%             Field_new( isnan(Field_new))=0;
            wbeam_out = beam(wbeam.X,wbeam.Y,Field_new,wbeam.wavelength);                 % 生成新的光束                              ；
        end
    end
end

File: src/optical_element/w_oe.cpp
#include "w_oe.h"

using namespace Furion_NS;

W_Oe::W_Oe(Beam* beam_in, double ds, double di, double chi, double theta, No_Surfe* surface, Grating* grating)
    : beam_in(beam_in), grating(grating), surface(surface), theta(theta), chi(chi), beam_out(nullptr), Cff(0), N(beam_in->N), ds(ds), di(di), N2(N*N)
{
    cout << "W_Oe 初始化" << endl;

}

W_Oe::~W_Oe()
{
    delete beam_out;
    delete beam_in;
    delete gbeam_in;
    delete grating;
    delete surface;
    //delete g_mirror;
    cout << "W_Oe 的析构" << endl;
}

void W_Oe::reflect(Beam* beam_in, double ds, double di, double chi, double theta)
{
    cout << "W_Oe de reflect" << endl;

    create_g_beam();

    string mirror = tracing();

    double* delta_h = new double[this->N2];
    double* phase_s = new double[this->N2];

    if (mirror == "Furion_Plane_Mirror")
    {
        this->surface->value(delta_h, this->g_mirror->Z2, this->g_mirror->X2, N);
        for (int i = 0; i < this->N2; i++)
        {
            phase_s[i] = -2 * 2 * _Pi / this->beam_in->wavelength * delta_h[i] * g_mirror->cos_Alpha[i];
        }
    }
    else if (mirror == "Furion_Cylinder_Ellipse_Mirror")
    {
        this->surface->value(delta_h, this->g_FCE_mirror->Z2, this->g_FCE_mirror->X2, N*N);
        for (int i = 0; i < this->N2; i++)
        {
            phase_s[i] = -2 * 2 * _Pi / this->beam_in->wavelength * delta_h[i] * g_FCE_mirror->cos_Alpha[i];
            //cout << phase_s[i] << endl;
        }
        //cout << endl;
    }

    create_w_beam(phase_s);
}

std::string W_Oe::tracing()
{
    cout << "G_Furion_ellipsoid_MirrorW_Oe的tracing" << endl;

    return ("w_oe");
}

void W_Oe::create_g_beam()
{
    cout << "W_Oe de create_g_beam" << endl;

    double** Phase_field;
    create_2d(Phase_field, N);
    f_p_u2.Furion_phase_unwrap2(Phase_field, this->beam_in->field, N);
    
    double** optical_path;
    create_2d(optical_path, N);
    create_2d(beam_in->phase_field, N);
    double Pi2 = 2 * _Pi;
    for (int i = 0; i < N; i++) 
    { 
        optical_path[i] = new double[N];
        for (int j = 0; j < N; j++)
        {
            optical_path[i][j] = Phase_field[i][j] / (Pi2)*beam_in->wavelength;
            beam_in->phase_field[i][j] = Phase_field[i][j];
            //cout << Phase_field[i][j] << " ";
        }
        //cout << endl;
    }

    destory_2d(Phase_field, N);

    double dx = mean_diff(beam_in->X, N, 0);
    double dy = mean_diff(beam_in->Y, N, 1);

    double** Fx;
    double** Fy;
    create_2d(Fx, N);
    create_2d(Fy, N);
    gradient(Fx, Fy, optical_path, dx, dy, N);
    destory_2d(optical_path, N);

    create_1d(beam_in->XX, this->N2);
    create_1d(beam_in->YY, this->N2);
    double* phi = new double[this->N2];
    double* psi = new double[this->N2];
    for (int i = 0; i < N; i++)
    {
        for (int j = 0; j < N; j++)
        {
            beam_in->XX[i + N * j] = beam_in->X[i][j];
            beam_in->YY[i + N * j] = beam_in->Y[i][j];
            phi[i + N * j] = -atan(Fx[i][j]);
            psi[i + N * j] = -atan(Fy[i][j]);
        }
    }

    destory_2d(Fx, N);
    destory_2d(Fy, N);

    gbeam_in = new G_Beam(beam_in->XX, beam_in->YY, phi, psi, beam_in->wavelength, this->N2);
}

void W_Oe::create_w_beam(double* s_phase)
{
    cout << "W_Oe de create_g_beam" << endl;

    //double** Phase;
    //create_2d(Phase, N);
    //f_p_u2.Furion_phase_unwrap2(Phase, this->beam_in->field, N);

    this->beam_out = new Beam(this->beam_in->X, this->beam_in->Y, this->beam_in->field, this->beam_in->wavelength, N);

}

double W_Oe::mean_diff(double** X, int N, int n)
{
    double dx = 0;

    if (n == 0)
    {
        double dx = 0;
        for (int i = 1; i < N; i++)
        {
            dx = dx + (X[0][i] - X[0][i - 1]);
        }
        return dx / (N - 1);
    }
    else if (n == 1)
    {
        double dx = 0;
        for (int i = 1; i < N; i++)
        {
            dx = dx + (X[i][0] - X[i - 1][0]);
        }
        return dx / (N-1);
    }
    else
    {
        cout << "The mean diff function can only accept 0 and 1 as arguments" << endl;
        exit(0);
    }
}

void W_Oe::gradient(double** Fx, double** Fy, double** optical_path, double dx, double dy, int N) 
{
    // Fx
    for (int i = 0; i < N; i++) {
        for (int j = 0; j < N; j++) {
            if (j == 0) {
                Fx[i][j] = (optical_path[i][j + 1] - optical_path[i][j]) / dx;
            }
            else if (j == N - 1) {
                Fx[i][j] = (optical_path[i][j] - optical_path[i][j - 1]) / dx;
            }
            else {
                Fx[i][j] = (optical_path[i][j + 1] - optical_path[i][j - 1]) / (2 * dx);
            }
        }
    }

    // Fy
    for (int i = 0; i < N; i++) {
        for (int j = 0; j < N; j++) {
            if (i == 0) {
                Fy[i][j] = (optical_path[i + 1][j] - optical_path[i][j]) / dy;
            }
            else if (i == N - 1) {
                Fy[i][j] = (optical_path[i][j] - optical_path[i - 1][j]) / dy;
            }
            else {
                Fy[i][j] = (optical_path[i + 1][j] - optical_path[i - 1][j]) / (2 * dy);
            }
        }
    }
}
